### PR TITLE
Remove AsyncLLMEngine busy loop, shield background task

### PR DESCRIPTION
**Pull request by Yard1**
This PR reduces CPU utilisation of `AsyncLLMEngine` by replacing busy looping with an `asyncio.Event` triggered whenever there are new requests, and makes sure that the background loop task is shielded from cancellation.

### CODE REVIEW

**Comment by Yard1** [comment]
cc @zhuohan123

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import asyncio
 import time
 from functools import partial
-from typing import Any, Dict, Iterable, List, Optional, Set, Tuple, Type, Union
+from typing import Any, Callable, Coroutine, Dict, Iterable, List, Optional, Set, Tuple, Type, Union
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:4]: Nit: add parenthesis around this long import and then run the formatter.
```suggestion
from typing import (Any, Callable, Coroutine, Dict, Iterable, List, Optional, Set, Tuple, Type, Union)
```

File: vllm/engine/async_llm_engine.py [hunk 2]
```diff
@@ -70,6 +70,40 @@ async def __anext__(self) -> RequestOutput:
...
+        self.task_set = set()
+
+    def run(self, coroutine: Coroutine, callback: Callable):
+        # Wrap the coroutine in a task so it executes immediately
+        # Save a reference to the task so it doesn't get garbage collected
+        # Remove the task from the task set when it is completed
+        task = asyncio.get_event_loop().create_task(coroutine)
+        self.task_set.add(task)
+        if callback:
+
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:92]: Nit: redundant new line
```suggestion
```

File: vllm/engine/async_llm_engine.py [hunk 3]
```diff
@@ -264,11 +317,11 @@ def start_background_loop(self) -> None:
         """Start the background loop."""
         if self.is_running:
             raise RuntimeError("Background loop is already running.")
-        self.background_loop = asyncio.get_event_loop().create_task(
-            self.run_engine_loop())
-        self.background_loop.add_done_callback(
-            partial(_raise_exception_on_finish,
-                    request_tracker=self.request_tracker))
+        self._request_tracker.init_event()
+        self.background_loop = self._task_set.run(
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:321]: if this `task_set` will only have one task, why is a "set" needed? Can we just shield the background loop?

File: vllm/engine/async_llm_engine.py [hunk 4]
```diff
@@ -304,18 +359,24 @@ async def engine_step(self):
 
         # Put the outputs into the corresponding streams.
         for request_output in request_outputs:
-            self.request_tracker.process_request_output(
+            self._request_tracker.process_request_output(
                 request_output, verbose=self.log_requests)
 
+        return bool(request_outputs)
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:365]: Nit: Make the code a bit easier to understand:
```suggestion
        return len(request_outputs) > 0
```

**Review by zhuohan123 — CHANGES_REQUESTED** [review]

File: vllm/engine/async_llm_engine.py (hunk 2 above)
Comment (Yard1) [reply vllm/engine/async_llm_engine.py:92]: added by formattter

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the fix!

**Pull request merged**